### PR TITLE
Allow for wildcards in k8s ingress host, fixes #792

### LINKED ISSUE #792: Question: Wildcard Host for Kubernetes Ingress
How do you create a wildcard host using kubernetes ingress? The following format does not work:

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: 3-0-proxy
spec:
  rules:
  - host: '*.mydomain.com'
    http:
      paths:
      - backend:
          serviceName: service-3-0-proxy
          servicePort: 80
        path: /
```

### PR DESCRIPTION
**Pull request by sheerun**
Example ingress:

```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  namespace: deis
  name: deis
spec:
  rules:
  - host: "*.example.com"
    http:
      paths:
      - path: /
        backend:
          serviceName: deis-router
          servicePort: 80
```

### CODE REVIEW

File: provider/kubernetes.go [hunk 1]
```diff
@@ -125,9 +125,15 @@ func (provider *Kubernetes) loadIngresses(k8sClient k8s.Client) (*types.Configur
 					}
 				}
 				if len(r.Host) > 0 {
+					rule := "Host:" + r.Host
+
+					if strings.Contains(r.Host, "*") {
+						rule = "HostRegexp:" + strings.Replace(r.Host, "*", "{subdomain:[a-z]+}", 1)
```
Comment (dtomcej) [inline provider/kubernetes.go]: The subdomain rule you implemented here does not allow for capital letters, numbers, or dashes/underscores. This means that `ww2.example.com` will not have a properly matched rule.

Comment (sheerun) [inline provider/kubernetes.go]: @dtomcej I've changed it to allow dashes and capital letter, but I believe underscores are disallowed in domain names, aren't they?

**Comment by dtomcej** [comment]
http://stackoverflow.com/a/2183140

The commenter is correct, many services include underscores in the subdomains.

**Comment by sheerun** [comment]
@dtomcej fixed

**Comment by dtomcej** [comment]
Why not just `"{subdomain:[A-Za-z0-9-_]*}"`? The underscore may be the first character in the subdomain:

_test.example.com

**Comment by sheerun** [comment]
@dtomcej  fixed

**Comment by dtomcej** [comment]
LGTM :shipit:
Ping @containous/traefik

**Comment by emilevauge** [comment]
Should we allow `_` ?
From [Wikipedia](https://en.wikipedia.org/wiki/Domain_name):

>Domain names may be formed from the set of alphanumeric ASCII characters (a-z, A-Z, 0-9), but characters are case-insensitive. In addition the hyphen is permitted if it is surrounded by characters, digits or hyphens, although it is not to start or end a label. Labels are always separated by the full stop (period) character in the textual name representation. 

We may keep the same regex `\\da-z\\.-`: https://github.com/containous/traefik/blob/master/server.go#L754

**Pull request merged**